File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.RemoteProxyExample/RemotePaymentServer.cs
namespace ProxyDesignPattern.RemoteProxyExample;

/// <summary>
/// Simulates a remote payment server running on a different machine.
/// In real scenarios, this would be an actual remote service.
/// </summary>
public class RemotePaymentServer
{
    private readonly Dictionary<string, decimal> _accounts = new()
    {
        { "ACC-001", 5000.00m },
        { "ACC-002", 1500.00m },
        { "ACC-003", 250.00m }
    };

    public string HandleRequest(string requestType, Dictionary<string, string> parameters)
    {
        Console.WriteLine($"[RemoteServer] Received request: {requestType}");
        Thread.Sleep(500); // Simulate network latency

        return requestType switch
        {
            "PROCESS_PAYMENT" => ProcessPaymentRequest(parameters),
            "GET_BALANCE" => GetBalanceRequest(parameters),
            _ => "ERROR|Unknown request type"
        };
    }

    private string ProcessPaymentRequest(Dictionary<string, string> parameters)
    {
        var amount = decimal.Parse(parameters["amount"]);
        var cardNumber = parameters["cardNumber"];
        
        Console.WriteLine($"[RemoteServer] Processing payment: {amount:C}");
        Thread.Sleep(1000); // Simulate processing time

        var transactionId = $"TXN-{Guid.NewGuid().ToString()[..8].ToUpper()}";
        
        Console.WriteLine($"[RemoteServer] Payment approved. Transaction: {transactionId}");
        return $"SUCCESS|{transactionId}|Payment processed successfully";
    }

    private string GetBalanceRequest(Dictionary<string, string> parameters)
    {
        var accountId = parameters["accountId"];
        
        if (_accounts.TryGetValue(accountId, out var balance))
        {
            Console.WriteLine($"[RemoteServer] Balance for {accountId}: {balance:C}");
            return $"SUCCESS|{balance}";
        }

        return "ERROR|Account not found";
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.RemoteProxyExample/src/RemotePaymentServer.ts
const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Simulates a remote payment server running on a different machine.
 * In real scenarios, this would be an actual remote service.
 */
export class RemotePaymentServer {
  private readonly _accounts: Map<string, number> = new Map([
    ["ACC-001", 5000.0],
    ["ACC-002", 1500.0],
    ["ACC-003", 250.0],
  ]);

  async handleRequest(
    requestType: string,
    parameters: Record<string, string>
  ): Promise<string> {
    console.log(`[RemoteServer] Received request: ${requestType}`);
    await sleep(500); // Simulate network latency

    switch (requestType) {
      case "PROCESS_PAYMENT":
        return this._processPaymentRequest(parameters);
      case "GET_BALANCE":
        return this._getBalanceRequest(parameters);
      default:
        return "ERROR|Unknown request type";
    }
  }

  private async _processPaymentRequest(
    parameters: Record<string, string>
  ): Promise<string> {
    const amount = parseFloat(parameters["amount"]);
    console.log(`[RemoteServer] Processing payment: $${amount.toFixed(2)}`);
    await sleep(1000); // Simulate processing time

    const transactionId = `TXN-${Math.random().toString(36).substring(2, 10).toUpperCase()}`;
    console.log(`[RemoteServer] Payment approved. Transaction: ${transactionId}`);
    return `SUCCESS|${transactionId}|Payment processed successfully`;
  }

  private _getBalanceRequest(parameters: Record<string, string>): string {
    const accountId = parameters["accountId"];
    const balance = this._accounts.get(accountId);

    if (balance !== undefined) {
      console.log(`[RemoteServer] Balance for ${accountId}: $${balance.toFixed(2)}`);
      return `SUCCESS|${balance}`;
    }

    return "ERROR|Account not found";
  }
}
